### PR TITLE
LUCENE-10196: Improve IntroSorter with 3-ways partitioning.

### PR DESCRIPTION
**Pull request by bruno-roustant**

### CODE REVIEW

**Review by dweiss — APPROVED** [review]
I wonder whether this will show up as a difference in real-life benchmarks!

**Review by jpountz — APPROVED** [review]
This is pretty cool! I tried to replace some of the radix sorter we're using with this improved IntroSorter to see whether they still performed better. They did, but maybe we need to revisit when to fall back from radix sort to intro sort in the future.

File: lucene/core/src/java/org/apache/lucene/util/IntroSorter.java [hunk 1]
```diff
@@ -20,66 +20,120 @@
...
+      // Pivot selection based on medians.
+      int last = to - 1;
+      int mid = (from + last) >>> 1;
+      int range = size >> 3;
+      int pivot;
+      if (size <= SINGLE_MEDIAN_THRESHOLD) {
+        // Select the pivot with a single median around the middle element.
+        // Do not take the median between [from, mid, last] because it hurts performance
+        // if the order is descending.
+        pivot = median(mid - range, mid, mid + range);
```
[the remark this answers is not in the document]
Comment (bruno-roustant) [reply lucene/core/src/java/org/apache/lucene/util/IntroSorter.java:69]: Ok, it is more intuitive.

File: lucene/core/src/java/org/apache/lucene/util/IntroSorter.java [hunk 2]
```diff
@@ -20,66 +20,120 @@
...
-        --right;
+      // Pivot selection based on medians.
+      int last = to - 1;
+      int mid = (from + last) >>> 1;
+      int range = size >> 3;
+      int pivot;
+      if (size <= SINGLE_MEDIAN_THRESHOLD) {
+        // Select the pivot with a single median around the middle element.
+        // Do not take the median between [from, mid, last] because it hurts performance
+        // if the order is descending.
```
[the remark this answers is not in the document]
Comment (bruno-roustant) [reply lucene/core/src/java/org/apache/lucene/util/IntroSorter.java:68]: It is counter-intuitive indeed. It is due to the way the 3-way partitioning algorithm swaps elements. At the beginning I used a median between [from, mid, last] but on a descending ordering and after a couple recursive levels, systematically the median was 'last' and it resulted in a bad partition with an empty right set, triggering a worst case. By taking this new median, we avoid the worst case whatever the order is.

File: lucene/core/src/java/org/apache/lucene/util/IntroSorter.java [hunk 3]
```diff
@@ -20,66 +20,120 @@
...
+        swap(k++, j--);
+      }
+      for (int k = last; k > q; ) {
+        swap(k--, i++);
       }
 
-      if (left < right) {
-        swap(left, right);
-        --right;
+      // Recursion on the smallest partition. Replace the tail recursion by a loop.
```
[the remark this answers is not in the document]
Comment (bruno-roustant) [reply lucene/core/src/java/org/apache/lucene/util/IntroSorter.java:112]: No, maybe I'm not clear. I replaced the two recursive calls on both left and right parts, by a recursive call on the smallest part and a loop on the largest part.

File: lucene/core/src/java/org/apache/lucene/util/IntroSorter.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (jpountz) [reply lucene/core/src/java/org/apache/lucene/util/IntroSorter.java:68]: Thanks for explaining.

File: lucene/core/src/java/org/apache/lucene/util/IntroSorter.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (jpountz) [reply lucene/core/src/java/org/apache/lucene/util/IntroSorter.java:112]: Ah, I had misunderstood what the comment meant indeed, thanks for explaining.

**Review by jpountz — APPROVED** [review]

**Pull request closed without merging**